src/acme.rs

```diff
@@ -487,11 +489,69 @@ where
 
         let certificate = order
             .certificate
-            .ok_or(NewCertificateError::CertificateUrl)?;
+            .ok_or(NewCertificateError::MissingCertificate)?;
 
-        let chain = self.post(&certificate, b"").await?.into_body();
+        let res = self.post(&certificate, b"").await?;
 
-        Ok(NewCertificateOutput { chain, pkey })
+        if let Some(ref matcher) = self.issuer.chain {
+            let (bytes, x509) = self
+                .find_preferred_chain(&certificate, res, matcher)
+                .await?;
+            Ok(NewCertificateOutput { bytes, x509, pkey })
+        } else {
+            let bytes = res.into_body();
+            let x509 =
+                X509::stack_from_pem(&bytes).map_err(NewCertificateError::InvalidCertificate)?;
+            if x509.is_empty() {
+                return Err(NewCertificateError::MissingCertificate);
+            }
+
+            Ok(NewCertificateOutput { bytes, x509, pkey })
+        }
+    }
+
+    async fn find_preferred_chain(
+        &self,
+        base: &Uri,
+        cert: http::Response<Bytes>,
+        matcher: &CertificateChainMatcher,
+    ) -> Result<(Bytes, std::vec::Vec<X509>), NewCertificateError> {
+        let default =
+            X509::stack_from_pem(cert.body()).map_err(NewCertificateError::InvalidCertificate)?;
```

Comment:
Arguably, it *shouldn't* happen but would it make sense to delay the `map_err` until the time where `default` comes the chosen chain? Proactively doing `map_err` here might result in errors even when admins have specifically configured to use a different chain (maybe for this reason?).